### ISSUE TITLE
Allow @MappingTarget to be used with @SubclassMapping

### ISSUE THREAD

**Issue by heanssgen-troy** [mapstruct/mapstruct#2946]
Currently, it is an error to use @MappingTarget in conjunction with @SubclassMapping.

This leads to the scenario where we can convert from a DTO into a subclass structure correctly, but cannot _update_ that subclass structure using another DTO. This is particularly annoying in a nested structure (such as queries and subqueries).

The current workaround is to write each subclass as an Update method and then use an @AfterMapping method to do the instanceof checks and call the appropriate update methods in the subclass. This is not ideal, as 'inherited' properties from the superclass will be mapped twice (causing excess database reads)/

It would be immensely helpful if the update method could inherit the @SubclassMapping annotations and perform the instanceof checks directly in the mapper. In the even the mapper could not find an appropriate subclass type, either throwing an exception or downcasting and using the current functionality would be good.

**Comment by filiphr** [comment]
This does sound like an interesting enhancements @heanssgen-troy. Would you be willing to work on adding this to the project?

**Comment by filiphr** [comment]
As I've said in https://github.com/mapstruct/mapstruct/issues/2946#issuecomment-1201504305, this is a good enhancement. We are busy with some other issues. Are you @jonsalvas perhaps interested in adding this functionality?

**Comment by Errore418** [comment]
I agree that this is a big limitation of the SubclassMapping annotation. I'm willing to try to implement it. Do you @filiphr have some tips to begin with? This is my first time contributing to this project.

**Comment by csabasulyok** [comment]
Would also be interested in this feature.

**Comment by D3PSI** [comment]
> Would also be interested in this feature.

same here, this would be really useful

**Comment by D3PSI** [comment]
i'd be willing to give implementing this myself a shot, if someone could provide guidance on what would need to be done in order to achieve this

**Comment by filiphr** [comment]
Thanks @D3PSI. 

The best place to start would be in the [BeanMappingMethod#createSubclassMapping](https://github.com/mapstruct/mapstruct/blob/03563d8ffea0ca5ff4eadb297795f2407001c5dd/processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java#L398). Have a look at how we create the subclass mappings for normal methods. There is also `SubclassMappingOptions#isConsistent` where we actually add the compiler error that we do not support them on update methods.

Have a look at `SubclassMappingTest` for the different test scenarios that we have.

Please let us know if you are still interested in working on this and I'd assign this issue to you

**Comment by D3PSI** [comment]
hi @filiphr, thanks for your response.

you may go ahead and assign the task to me. I can't however promise that I'll get to it in a timely manner, but of course I'll stand by my word and see what I can do here.

cheers !

**Comment by filiphr** [comment]
@D3PSI take all the time you need. If you have any questions feel free to reach out.

**Comment by s-jepsen** [comment]
Any news on this issue?

**Comment by D3PSI** [comment]
i have started to work on this last week, no news so far. i will try to open a PR that adresses this ticket in due time

**Comment by D3PSI** [comment]
@filiphr i have now opened a PR that adresses this feature request here: #3330

please take a look and let me know if there's something to improve still! thank you all.

**Comment by D3PSI** [comment]
(also, does someone know how to properly escape `#` characters in git commit messages? i'd appreciate it, lol)

**Comment by filiphr** [comment]
Thanks for your work on this @D3PSI, I've left some comments on the PR. Have a look at my second remark, I am not sure how we are going to solve this properly

**Comment by Caillak** [comment]
Hello !

Any chance this PR move forward ?
